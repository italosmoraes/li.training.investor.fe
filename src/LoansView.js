import React, {Component} from 'react';
import LoanCard from './LoanCard';
import InvestView from './InvestView';
import loansList from './LoansList';

class LoansView extends Component {

    constructor(props){
        super(props);
        this.state = {loans: this.transformInputData(loansList.loans)};
    }

    transformInputData = (loans) => {
        return loans.map( (loan) => Object.assign({}, loan, {
            available: +(loan.available.replace(',', '')),
            expiration_date: new Date((new Date()).getTime() + loan.term_remaining * 1000)
        }) )
    }

    updateAvailableAmount = (loanId, investment) => {
        let loans = this.state.loans;
        let newLoans = loans.reduce((acc,loan) => {
            if(loan.id === loanId){
                let newLoan = Object.assign({}, loan, {
                    available: loan.available - investment
                });
                return [...acc, newLoan]
            }
            return [...acc, loan];
        }, [])
        this.setState({loans: newLoans});
    }

    

    render(){
        let loans = this.state.loans.map(loan => 
            <LoanCard 
                key={loan.id} 
                loan={loan} 
                updateAvailableAmount={this.updateAvailableAmount}
            /> 
        );
        return(
            <div class="loans-view">
                <h2 id="loan-view-title"> Current Loans </h2>
                {loans}
            </div>
        )
    }

}

export default LoansView;