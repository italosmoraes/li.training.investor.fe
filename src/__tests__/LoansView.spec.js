import React from 'react';
import LoansView from '../LoansView';
import { shallow } from 'enzyme';

describe('LoansView', () => {
    let wrapper;
    let instance;

    beforeEach(() => {
        wrapper = shallow(<LoansView/>)
        instance = wrapper.instance();
    });

    describe('updateAvailableAmount', () => {
        it('update states with new Amount for given loan', () => {
            instance.state = {
                loans: [{id: 5003, available: 300}, {id: 10, available: 500}]
            }
            instance.setState = jest.fn();
            instance.updateAvailableAmount(0,10);
            expect(instance.setState).toHaveBeenCalledWith({
                loans: [{id: 5003, available: 300}, {id: 10, available: 500}]
            });
            instance.setState = jest.fn();
            instance.updateAvailableAmount(5003,200);
            expect(instance.setState).toHaveBeenCalledWith({
                loans: [{id: 5003, available: 100}, {id: 10, available: 500}]
            })
            instance.setState = jest.fn();
            instance.updateAvailableAmount(10,500);
            expect(instance.setState).toHaveBeenCalledWith({
                loans: [{id: 5003, available: 300}, {id: 10, available: 0}]
            })
            
        })
    })
    describe('transformInputData', () => {
        it('transform available amount from text to number', () => {
            expect(instance.transformInputData([{
                available: '19,888.43', 
                term_remaining: 0}])
            ).toEqual([{
                available: 19888.43,
                expiration_date: expect.any(Date),
                term_remaining: 0
            }])
        })
    })
    describe('transformInputData', () => {
        it('transform text into Date', () => {
            const constantDate = new Date('2017-06-13T04:41:20')
            const constantTimestampt = constantDate.getTime();

            /*eslint no-global-assign:off*/
            Date = class extends Date {
                getTime() {
                    return constantTimestampt;
                }
            }
            expect(instance.transformInputData([{
                available: '0', 
                term_remaining: '60'}])
            ).toEqual([{
                available: 0,
                expiration_date: new Date('2017-06-13T04:42:20'),
                term_remaining: '60'
            }])
        })
    })
})