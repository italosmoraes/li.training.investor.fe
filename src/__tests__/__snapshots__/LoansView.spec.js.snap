// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoansView LoanView is rendered renders the loans view 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoansView />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          id="loan-view-title"
        >
           Current Loans 
        </h2>,
        Array [
          <LoanCard
            loan={
              Object {
                "amount": "85,754",
                "annualised_return": "8.60",
                "available": 11959,
                "expiration_date": 2017-06-23T04:41:20.000Z,
                "id": "1",
                "ltv": "48.80",
                "term_remaining": "864000",
                "title": "Voluptate et sed tempora qui quisquam.",
                "tranche": "A",
              }
            }
            updateAvailableAmount={[Function]}
          />,
          <LoanCard
            loan={
              Object {
                "amount": "85,754",
                "annualised_return": "7.10",
                "available": 31405,
                "expiration_date": 2017-07-01T22:41:20.000Z,
                "id": "5",
                "ltv": "48.80",
                "term_remaining": "1620000",
                "title": "Consectetur ipsam qui magnam minus dolore ut fugit.",
                "tranche": "B",
              }
            }
            updateAvailableAmount={[Function]}
          />,
          <LoanCard
            loan={
              Object {
                "amount": "85,754",
                "annualised_return": "4.80",
                "available": 12359,
                "expiration_date": 2017-06-23T08:51:20.000Z,
                "id": "12",
                "ltv": "48.80",
                "term_remaining": "879000",
                "title": "Dolores repudiandae ut voluptas unde laborum quaerat et sapiente.",
                "tranche": "C",
              }
            }
            updateAvailableAmount={[Function]}
          />,
        ],
      ],
      "class": "loans-view",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " Current Loans ",
          "id": "loan-view-title",
        },
        "ref": null,
        "rendered": " Current Loans ",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "loan": Object {
            "amount": "85,754",
            "annualised_return": "8.60",
            "available": 11959,
            "expiration_date": 2017-06-23T04:41:20.000Z,
            "id": "1",
            "ltv": "48.80",
            "term_remaining": "864000",
            "title": "Voluptate et sed tempora qui quisquam.",
            "tranche": "A",
          },
          "updateAvailableAmount": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "5",
        "nodeType": "class",
        "props": Object {
          "loan": Object {
            "amount": "85,754",
            "annualised_return": "7.10",
            "available": 31405,
            "expiration_date": 2017-07-01T22:41:20.000Z,
            "id": "5",
            "ltv": "48.80",
            "term_remaining": "1620000",
            "title": "Consectetur ipsam qui magnam minus dolore ut fugit.",
            "tranche": "B",
          },
          "updateAvailableAmount": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "12",
        "nodeType": "class",
        "props": Object {
          "loan": Object {
            "amount": "85,754",
            "annualised_return": "4.80",
            "available": 12359,
            "expiration_date": 2017-06-23T08:51:20.000Z,
            "id": "12",
            "ltv": "48.80",
            "term_remaining": "879000",
            "title": "Dolores repudiandae ut voluptas unde laborum quaerat et sapiente.",
            "tranche": "C",
          },
          "updateAvailableAmount": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            id="loan-view-title"
          >
             Current Loans 
          </h2>,
          Array [
            <LoanCard
              loan={
                Object {
                  "amount": "85,754",
                  "annualised_return": "8.60",
                  "available": 11959,
                  "expiration_date": 2017-06-23T04:41:20.000Z,
                  "id": "1",
                  "ltv": "48.80",
                  "term_remaining": "864000",
                  "title": "Voluptate et sed tempora qui quisquam.",
                  "tranche": "A",
                }
              }
              updateAvailableAmount={[Function]}
            />,
            <LoanCard
              loan={
                Object {
                  "amount": "85,754",
                  "annualised_return": "7.10",
                  "available": 31405,
                  "expiration_date": 2017-07-01T22:41:20.000Z,
                  "id": "5",
                  "ltv": "48.80",
                  "term_remaining": "1620000",
                  "title": "Consectetur ipsam qui magnam minus dolore ut fugit.",
                  "tranche": "B",
                }
              }
              updateAvailableAmount={[Function]}
            />,
            <LoanCard
              loan={
                Object {
                  "amount": "85,754",
                  "annualised_return": "4.80",
                  "available": 12359,
                  "expiration_date": 2017-06-23T08:51:20.000Z,
                  "id": "12",
                  "ltv": "48.80",
                  "term_remaining": "879000",
                  "title": "Dolores repudiandae ut voluptas unde laborum quaerat et sapiente.",
                  "tranche": "C",
                }
              }
              updateAvailableAmount={[Function]}
            />,
          ],
        ],
        "class": "loans-view",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " Current Loans ",
            "id": "loan-view-title",
          },
          "ref": null,
          "rendered": " Current Loans ",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "loan": Object {
              "amount": "85,754",
              "annualised_return": "8.60",
              "available": 11959,
              "expiration_date": 2017-06-23T04:41:20.000Z,
              "id": "1",
              "ltv": "48.80",
              "term_remaining": "864000",
              "title": "Voluptate et sed tempora qui quisquam.",
              "tranche": "A",
            },
            "updateAvailableAmount": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "5",
          "nodeType": "class",
          "props": Object {
            "loan": Object {
              "amount": "85,754",
              "annualised_return": "7.10",
              "available": 31405,
              "expiration_date": 2017-07-01T22:41:20.000Z,
              "id": "5",
              "ltv": "48.80",
              "term_remaining": "1620000",
              "title": "Consectetur ipsam qui magnam minus dolore ut fugit.",
              "tranche": "B",
            },
            "updateAvailableAmount": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "12",
          "nodeType": "class",
          "props": Object {
            "loan": Object {
              "amount": "85,754",
              "annualised_return": "4.80",
              "available": 12359,
              "expiration_date": 2017-06-23T08:51:20.000Z,
              "id": "12",
              "ltv": "48.80",
              "term_remaining": "879000",
              "title": "Dolores repudiandae ut voluptas unde laborum quaerat et sapiente.",
              "tranche": "C",
            },
            "updateAvailableAmount": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
